### ISSUE TITLE
Treating `user` and `object` type-only values inconsistently when reading tuples in postgres

### ISSUE THREAD

**Issue by iamolegga** [openfga/openfga#1940]
### Checklist

- [X] I have looked into the [README](https://github.com/openfga/openfga/blob/main/README.md) and have not found a suitable solution or answer.
- [X] I have looked into the [documentation](https://openfga.dev/docs) and have not found a suitable solution or answer.
- [X] I have searched the [issues](https://github.com/openfga/openfga/issues) and have not found a suitable solution or answer.
- [X] I have upgraded to the [latest version](https://github.com/openfga/openfga/releases/latest) of OpenFGA and the issue still persists.
- [X] I have searched the [OpenFGA Community](https://openfga.dev/community) and have not found a suitable solution or answer.
- [X] I agree to the terms within the [OpenFGA Code of Conduct](https://github.com/openfga/.github/blob/main/CODE_OF_CONDUCT.md).

### Description

I don't know if this is a bug or a feature, but it looks strange and unexpected.

When getting tuples passing full user (type+id) and type only for the object it works:

```sh
λ  builder main ✓ fga tuple read --store-id=01J82JPK1AVKSJQJAK35DNWWNB --user user:ybwtv2146mir --object organization:
{
  "continuation_token":"",
  "tuples": [
    {
      "key": {
        "object":"organization:b7851d2d-79a8-4c07-915d-286f1b308716",
        "relation":"owner",
        "user":"user:ybwtv2146mir"
      },
      "timestamp":"2024-09-18T14:37:15.560864Z"
    }
  ]
}
```

When doing the same the opposite way (type only for the user and full for the object) it doesn't:

```sh
λ  builder main ✓ fga tuple read --store-id=01J82JPK1AVKSJQJAK35DNWWNB --user user: --object organization:b7851d2d-79a8-4c07-915d-286f1b308716
{
  "continuation_token":"",
  "tuples": []
}
```

Omitting the `user` field works fine:

```sh
λ  builder main ✓ fga tuple read --store-id=01J82JPK1AVKSJQJAK35DNWWNB --object organization:b7851d2d-79a8-4c07-915d-286f1b308716
{
  "continuation_token":"",
  "tuples": [
    {
      "key": {
        "object":"organization:b7851d2d-79a8-4c07-915d-286f1b308716",
        "relation":"owner",
        "user":"user:ybwtv2146mir"
      },
      "timestamp":"2024-09-18T14:37:15.560864Z"
    }
  ]
}
λ  builder main ✓
```

If I'm not mistaken, this behavior is due to this part of the code:

https://github.com/openfga/openfga/blob/main/pkg/storage/postgres/postgres.go#L195-L197

which is different from how `object` is handled:

https://github.com/openfga/openfga/blob/main/pkg/storage/postgres/postgres.go#L185-L191

### OpenFGA version

latest

### How are you running OpenFGA?

In Docker

### What datastore are you using?

Postgres

### OpenFGA Flags

no

**Comment by rhamzeh** [comment]
Hi @iamolegga - that is not supported, please see the supported Read queries here:

https://openfga.dev/api/service#/Relationship%20Tuples/Read

TLDR:


1.  `tuple_key` is optional. If not specified, it will return all tuples in the store.
2.  `tuple_key.object` is mandatory if tuple_key is specified. It can be a full object (e.g., `type:object_id`) or type only (e.g., `type:`).
3. `tuple_key.user` is mandatory if tuple_key is specified in the case the `tuple_key.object` is a type only.


The [Documentation in the SDK README](https://github.com/openfga/js-sdk?tab=readme-ov-file#read-relationship-tuples) might make this clearer
```js
// Find if a relationship tuple stating that a certain user is a viewer of a certain document
{
  user: "user:81684243-9356-4421-8fbf-a4f8d36aa31b",
  relation: "viewer",
  object: "document:roadmap",
};

// Find all relationship tuples where a certain user has any relation to a certain document
{
  user: "user:81684243-9356-4421-8fbf-a4f8d36aa31b",
  object: "document:roadmap",
};

// Find all relationship tuples where a certain user is a viewer of any document
{
  user: "user:81684243-9356-4421-8fbf-a4f8d36aa31b",
  relation: "viewer",
  object: "document:",
};

// Find all relationship tuples where a certain user has any relation with any document
{
  user: "user:81684243-9356-4421-8fbf-a4f8d36aa31b",
  object: "document:",
};

// Find all relationship tuples where any user has any relation with a particular document
{
  object: "document:roadmap",
};

// Read all stored relationship tuples
{};
```

Those are the only supported patterns

**Comment by miparnisari** [comment]
Looks like we need better validation around API inputs.

**Comment by iamolegga** [comment]
@rhamzeh thank you. I've been reading the docs all day long as I faced this unexpected (from my point of view) behavior. I call it _unexpected_ as I've seen the `tuple` table I don't understand the technical limitations of why we can't pass some type for the user and type+id for the object the same way that we can pass type for the object and type+id for user.

Let me explain the whole problem:

I have the next schema

```
type foo
  relations
    define bar: [bar]

bar
  relations
    define foo: [foo]
    define baz: [baz]

baz
  relations
    define bar: [bar]
```

When deleting `bar:bar_id` from my main database I delete all the related data automatically with `cascade`. But when deleting it from openfga I need to get all the tuples by setting `bar:bar_id` first as a user and second as an object, and after deleting all those tuples. Not too handy, but ok.

So I've started to get those tuples as `user`:
passed only `user: "bar:bar_id"` -> error, have to provide some other fileds, ok, let's add an object type `object: "foo:"` and `object: "baz:"` -> works

Now let's do the very same but passing `bar:bar_id` as `object`:
passing `object: "bar:bar_id"` and `user: "foo:"` and `user: "baz:"` -> no errors, but nothing returned. So, in that case, it could be omitted 😀 

So I was surprised by the API 3 times while doing that small feature:
1. no API to delete tuples with filters
2. not expected limitations for getting tuples by `user`, something more should be provided as well
3. empty result for getting tuples when providing `user:"type:" object:"type:id"` as object field is parsed underhood, but user field passed as is

Please treat my feedback as something that could be improved, as despite that, I really like this project, I've found its modeling language more convenient than in other projects. And I like also the robust ecosystem around it: docs, vscode extension, tests, visual playground, cli.

**Comment by aaguiarz** [comment]
Hi @iamolegga 

We have these two entries in our Roadmap related to the improvements you suggest:

https://github.com/openfga/roadmap/issues/33
https://github.com/openfga/roadmap/issues/34

I'll close this issue, please watch those for updates :) 

Thanks for the feedback!
